### PR TITLE
[eas-cli] Support for local builds

### PR DESCRIPTION
**Pull request by wkozyra95**
# Checklist

- [x] I've added an entry to [CHANGELOG.md](https://github.com/expo/eas-cli/blob/main/CHANGELOG.md#main) if necessary.

# Why

Support for local builds

# How

- run `eas-build-local`(this name might not be final version) executable and pass job object to stdin
- `eas-build-local` path is resolved by
  - checking EAS_LOCAL_BUILD_PLUGIN_PATH env (for development)
  - using node `require.resolve()` (binnaries installed globally via yarn often are not in PATH)
  - looking for binary in PATH

# Test Plan

run builds

### CODE REVIEW

File: packages/eas-cli/src/build/local.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+        `Install eas-build-local package from npm e.g. {chalk.bold(
+          'npm install -g eas-build-local'
+        )} and make sure that executable is in your PATH.`
+      );
+      throw err;
+    }
+  }
+}
+
+async function findWithNodeResoultion(): Promise<string | undefined> {
```
Comment (dsokal) [inline packages/eas-cli/src/build/local.ts:29]:
```suggestion
async function findWithNodeResolution(): Promise<string | undefined> {
```

File: packages/eas-cli/src/build/local.ts [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+  try {
+    const arg = Buffer.from(JSON.stringify({ job })).toString('base64');
+    await spawnAsync(execNameOrPath, [arg], {
+      stdio: 'inherit',
+    });
+  } catch (err) {
+    if (err.code === 'ENOENT') {
+      Log.warn(`Could not resolve executable ${execNameOrPath}.`);
+      Log.warn(
+        `Install eas-build-local package from npm e.g. {chalk.bold(
```
Comment (dsokal) [inline packages/eas-cli/src/build/local.ts:20]:
```suggestion
        `Install eas-build-local package from npm e.g. ${chalk.bold(
```

File: packages/eas-cli/src/build/local.ts [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+          'npm install -g eas-build-local'
+        )} and make sure that executable is in your PATH.`
```
Comment (dsokal) [inline packages/eas-cli/src/build/local.ts:22]:
```suggestion
        )} and make sure it's in PATH.`
```

File: packages/eas-cli/src/commands/build/index.ts [hunk 4]
```diff
@@ -52,6 +52,10 @@ export default class Build extends Command {
       default: false,
       description: 'Run command in --non-interactive mode',
     }),
+    local: flags.boolean({
+      default: false,
+      description: 'Run build locally.',
```
Comment (dsokal) [inline packages/eas-cli/src/commands/build/index.ts:57]:
```suggestion
      description: 'Run build locally',
```

File: packages/eas-cli/src/commands/build/index.ts [hunk 5]
```diff
@@ -76,12 +80,17 @@ export default class Build extends Command {
       const { exp } = getConfig(projectDir, { skipSDKVersionRequirement: true });
       const projectId = await getProjectIdAsync(exp);
 
-      if (!(await isEasEnabledForProjectAsync(projectId))) {
+      if (!flags.local || !(await isEasEnabledForProjectAsync(projectId))) {
```
Comment (dsokal) [inline packages/eas-cli/src/commands/build/index.ts:83]:
```suggestion
      if (!flags.local && !(await isEasEnabledForProjectAsync(projectId))) {
```

File: packages/eas-cli/src/commands/build/index.ts [hunk 6]
```diff
@@ -161,6 +171,17 @@ async function ensureNoPendingBuildsExistAsync({
   return false;
 }
 
+function verifyOptionsForLocalBuilds(platform: RequestedPlatform): boolean {
+  if (platform === RequestedPlatform.All) {
+    Log.error('Builds for multiple platforms are not supported with flag --local');
+    return false;
+  } else if (process.platform !== 'darwin' && platform === RequestedPlatform.iOS) {
+    Log.error('Unsupported platform, macOS is required to build apps for iOS.');
```
Comment (dsokal) [inline packages/eas-cli/src/commands/build/index.ts:179]:
```suggestion
    Log.error('Unsupported platform, macOS is required to build apps for iOS');
```

File: packages/eas-cli/src/commands/build/index.ts [hunk 7]
```diff
@@ -161,6 +171,17 @@ async function ensureNoPendingBuildsExistAsync({
...
+    return false;
+  }
+  return true;
```
Comment (dsokal) [inline packages/eas-cli/src/commands/build/index.ts:182]:
```suggestion
  } else {
    return true;
  }
```

File: packages/eas-cli/src/build/local.ts [hunk 8]
```diff
@@ -0,0 +1,40 @@
+import { Job } from '@expo/eas-build-job';
+import spawnAsync from '@expo/spawn-async';
+import fs from 'fs-extra';
+import path from 'path';
+
+import Log from '../log';
+
+export async function runLocalBuildAsync(job: Job): Promise<void> {
+  const execNameOrPath =
+    process.env.EAS_BUILD_LOCAL_PATH ?? (await findWithNodeResoultion()) ?? 'eas-build-local';
```
Comment (wkozyra95) [inline packages/eas-cli/src/build/local.ts:10]:
```suggestion
    process.env.EAS_BUILD_LOCAL_PATH ?? (await findWithNodeResolution()) ?? 'eas-build-local';
```

File: packages/eas-cli/src/build/local.ts [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+import fs from 'fs-extra';
+import path from 'path';
+
+import Log from '../log';
+
+export async function runLocalBuildAsync(job: Job): Promise<void> {
+  const execNameOrPath =
+    process.env.EAS_LOCAL_BUILD_PLUGIN_PATH ??
+    (await findWithNodeResolution()) ??
+    'eas-cli-local-build-plugin';
```
Comment (dsokal) [inline packages/eas-cli/src/build/local.ts:13]: It's "hardcoded" 4 times. Please assign this to a constant.

**Review by dsokal — APPROVED** [review]

**Pull request merged**